src/thread_task.rs

```diff
@@ -346,91 +435,61 @@ unsafe fn create_filter_sbrow(
-        if pf_t.is_some() && t.sby != 0 {
-            tasks[prev_t.unwrap()].next = pf_t;
-            prev_t = pf_t;
-            pf_t = None;
-        }
-        // re-borrow to avoid conflict with tasks[prev_t] above
-        let t = &mut tasks[t_idx];
-        t.recon_progress = 0 as c_int;
-        t.deblock_progress = 0 as c_int;
-        t.deps_skip = 0 as c_int;
-        t.type_0 = if pass != 1 {
-            TaskType::TileReconstruction
-        } else {
-            TaskType::TileEntropy
-        };
-        t.frame_idx = fc.index as c_uint;
-        if let Some(prev_t) = prev_t {
-            tasks[prev_t].next = Some(t_idx);
-        }
-        prev_t = Some(t_idx);
-        tile_idx += 1;
-    }
-    if pf_t.is_some() {
-        tasks[prev_t.unwrap()].next = pf_t;
-        prev_t = pf_t;
-    }
-    tasks[prev_t.unwrap()].next = None;
     fc.task_thread.done[(pass & 1) as usize].store(0, Ordering::SeqCst);
-    let mut pending_tasks = fc.task_thread.pending_tasks.lock().unwrap();
-    if !(pending_tasks.head.is_none() || pass == 2) {
-        unreachable!();
-    }
-    if pending_tasks.head.is_none() {
-        pending_tasks.head = tile_tasks;
-    } else {
-        tasks[pending_tasks.tail.unwrap()].next = tile_tasks;
+    create_filter_sbrow(fc, f, pass)?;
+    {
+        let mut pending_tasks = tasks.pending_tasks.lock();
+        for tile_idx in 0..num_tasks {
+            let ts = &f.ts[tile_idx];
+            let t = Rav1dTask {
+                sby: ts.tiling.row_start >> f.sb_shift,
+                recon_progress: 0,
+                deblock_progress: 0,
+                deps_skip: AtomicI32::new(0),
+                type_0: if pass != 1 {
+                    TaskType::TileReconstruction
+                } else {
+                    TaskType::TileEntropy
+                },
+                frame_idx: fc.index as c_uint,
+                tile_idx: tile_idx as c_uint,
+                next: Default::default(),
+            };
+            pending_tasks.push(t);
+        }
     }
-    pending_tasks.tail = prev_t;
-    fc.task_thread
-        .pending_tasks_merge
-        .store(1, Ordering::SeqCst);
+    tasks.pending_tasks_merge.store(true, Ordering::SeqCst);
     fc.task_thread.init_done.store(1, Ordering::SeqCst);
     Ok(())
 }
 
-pub(crate) unsafe fn rav1d_task_frame_init(c: &Rav1dContext, fc: &Rav1dFrameContext) {
+pub(crate) fn rav1d_task_frame_init(c: &Rav1dContext, fc: &Rav1dFrameContext) {
     fc.task_thread.init_done.store(0, Ordering::SeqCst);
-    let tasks = fc.task_thread.tasks();
-    let t_idx = Rav1dTaskIndex::Init;
-    let t = &mut (*tasks)[t_idx];
-    t.type_0 = TaskType::Init;
-    t.frame_idx = fc.index as c_uint;
-    t.sby = 0 as c_int;
-    t.deblock_progress = 0 as c_int;
-    t.recon_progress = t.deblock_progress;
-    insert_task(c, fc, t_idx, 1 as c_int);
+    let init_task = Rav1dTask::init(fc.index as c_uint);
+    fc.task_thread.insert_task(c, init_task, 1 as c_int);
```

Comment:
nit since we already removed most of these other redundant `as` casts.

```suggestion
    fc.task_thread.insert_task(c, init_task, 1);
```